nussl/core/mixing.py

```diff
@@ -3,11 +3,139 @@
+            auto: Automatically chooses direct or Fourier method based on an estimate of which is faster (default).
+        normalize: Whether to apply a normalization factor which will prevent clipping. Defaults to True.
+        scale: Whether to scale the output convolved signal to have the same max as the input, so they are of roughly equal loudness.
+    
+    Example:
+
+    >>> import nussl
+    >>> path = nussl.efz_utils.download_audio_file('schoolboy_fascination_excerpt.wav')
+    >>> sig = nussl.AudioSignal(path)
+    >>> path = nussl.efz_utils.download_audio_file('h072_Bar_2txts.wav')
+    >>> ir = nussl.AudioSignal(path)
+    >>> convolved = nussl.mixing.convolve(sig, ir)
+    >>> # Or apply with fluent interface
+    >>> convolved = sig.convolve(ir) 
+    
+    """
+    output = []
+    factor = 1.0
+
+    signal = copy.deepcopy(signal)
+    other = copy.deepcopy(other)
+
+    if normalize: factor = np.sum(np.abs(other.audio_data) ** 2)
+    
+    if signal.num_channels != 1 and other.num_channels != 1:
+        if signal.num_channels != other.num_channels:
+            raise RuntimeError(
+                "If both signals are multichannel, they must have the " 
+                "same number of channels!")
+        for s1_ch, s2_ch in zip(signal.get_channels(), other.get_channels()):
+            convolved_ch = scipy.signal.convolve(
+                s1_ch, s2_ch / factor, mode='full', method=method)
+            output.append(convolved_ch)
+    else:
+        for i, s1_ch in enumerate(signal.get_channels()):
+            for j, s2_ch in enumerate(other.get_channels()):
+                convolved_ch = scipy.signal.convolve(
+                    s1_ch, s2_ch / factor, mode='full', method=method)
+                output.append(convolved_ch)
+    
+    output = np.array(output)
+    if scale:
+        max_output = np.abs(output).max()
+        max_input = np.abs(signal.audio_data).max()
+        scale_factor = max_input / max_output
+        output *= scale_factor
+    
+    convolved_signal = signal.make_copy_with_audio_data(
+        output, verbose=False)
+    convolved_signal.truncate_samples(signal.signal_length)
+            
+    return convolved_signal
+
+def mix_audio_signals(fg_signal, bg_signal, snr=10):
+    """
+    Mixes noise with signal at specified 
+    signal-to-noise ratio. Returns the mix, and the altered foreground
+    and background sources.
+
+    Args:
+        fg_signal: AudioSignal object, will be the louder signal (or quieter if snr is negative).
+        bg_signal: AudioSignal object, will be the quieter signal (or louder if snr is negative).
+        snr: Signal-to-noise ratio in decibels to mix at.
+        inplace: Whether or not to do this operation in place.
+        return_sources: Whether to return the sources or return the mix.
+    """
+    fg_signal = copy.deepcopy(fg_signal)
+    bg_signal = copy.deepcopy(bg_signal)
+
+    pad_len = max(0, fg_signal.signal_length - bg_signal.signal_length)
```

Comment:
Two things:

1. I think you should bubble up whether to do `min(len(fg), len(bg))` and truncate the longer one or `max(len(fg), len(bg))` and pad the shorter one. (I don't think the former is useful, but I figure why not give the option to the user?) I think the opinionated way you've written it now is too subtle and could cause unintended side effects.
2. Should that be it's own subroutine? Not sure if it's much use outside this function, but maybe `def _match_signal_lengths(signal_list, mode='pad')` or even something on AudioSignal like `signal1.match_length(signal2)`. Just spitballing, lmk what you think

Reply:
btw: By "bubble up" I mean add an arg for the user.